### ISSUE TITLE
No UDP received from Server to Client on Android > 5.0

### ISSUE THREAD

**Issue by codeflakes0** [EsotericSoftware/kryonet#106]
I'm not sure the google group is still read so I'm reposting the issue here.
Please see this post "https://groups.google.com/forum/#!topic/kryonet-users/7l436tftBlo" for details.

Basically with the simplest Client / Server application you can't send UDP from Server
to Client on Android > 5.0. The same code works on Android 4.4.

UDP sending from Client to Server works though ...

PS: I've tested it on localhost so with Client and Server on the same device but this should not be a problem and the same code works on 4.4.

**Comment by NathanSweet** [comment]
It's still read, it's just I don't usually have time to setup and debug on
Android, which tends to have buggy networking with OS version and device
specific issues. Sorry I can't be of further help!

**Comment by codeflakes0** [comment]
Thank you for your answer Nathan. I've since checked as stated on the Google code forum that read() works and receive() returns null on android 5.0. I'm not familiar with the socket API and kryonet source code. Can I just use read() instead of receive() in readFromAdress() of UdpConnnection.java ?

**Comment by codeflakes0** [comment]
The following patch seems to work on my test app.

	public InetSocketAddress readFromAddress () throws IOException {
		DatagramChannel datagramChannel = this.datagramChannel;
		if (datagramChannel == null) throw new SocketException("Connection is closed.");
		lastCommunicationTime = System.currentTimeMillis();

        if(!datagramChannel.isConnected())
            return (InetSocketAddress)datagramChannel.receive(readBuffer); // always null on Android >= 5.0
        datagramChannel.read(readBuffer);
		return connectedAddress;
	}

**Comment by desertkun** [comment]
codeflakes0, thank you SO MUCH! I've wasted 2 days trying to get this issue down, and finally.

**Comment by selenir** [comment]
Okay guys, I have the same problem. I'm using Android Studio and somehow, it doesn't save the edited UDPConnection file. I made a post here: http://stackoverflow.com/questions/33335537/dark-blue-file-names-in-android-studio/33335876#33335876

Do you know what may cause this?

**Comment by desertkun** [comment]
I've forked kryonet to https://github.com/desertkun/kryonet and added this dirty fix.

1. Clone it
2. `mvn compile package -Dmaven.test.skip=true` (unfortunately, tests are failing)
3. `target/kryonet-2.22.0-RC2.jar`

or `mvn compile install` it into your local maven repo.

**Comment by selenir** [comment]
Thank you for your response desert!
So there is no way to edit already downloaded repo?

**Comment by desertkun** [comment]
You can just change remote `origin` to https://github.com/desertkun/kryonet.git and pull.
Well, I can give you a jar, but thats means you're completely alone with that file, and you can't do anything about it, so I would not recommend you to do that.

**Comment by selenir** [comment]
Thank you sir!!!

**Comment by selenir** [comment]
I'm sorry sir, but I can't make it working. I have pulled it once more, I have a folder with Kryonet, but I have set  this in gradle:

 compile "com.esotericsoftware:kryo:3.0.1"
    compile ("com.esotericsoftware:kryonet:2.22.0-RC1") {
        exclude module : 'kryo'
    }

I totally don't know how gradle works and it still reads the code from their repo somehow. Should I just remove this or something?

**Comment by desertkun** [comment]
Just change `com.esotericsoftware:kryonet:2.22.0-RC1` to `com.esotericsoftware:kryonet:2.22.0-RC2`. since second one is from my repository, and if you did install git repo into maven correctly, it will pull needed data.

Also, make sure you have `mavenLocal` record in your repositories section:

    repositories {
        mavenLocal()
        mavenCentral()
        ....
    }

Also, this discussion is sort of off topic, so please leave email or something, and I'll try help you in private.

**Comment by Magic-WoW** [comment]
My email is inszejn@gmail.com This is my second github account.
I need to download it via gradle. Already tried with forking your repo to my own, made a release and it still downloads normal Kryonet:

 compile "com.github.Magic-WoW:kryo:asd"
    compile ("com.github.Magic-WoW:kryonet:test"){
        exclude module : 'kryo'
    }

I just need to synch your Kryonet (with the fix) with my existing android studio project, that's why I can't pull it by VCS, I need to set it in gradle, like earlier. 

If this is possible, we can delete posts here. Just do it, if it's possible, I'll do the same.

**Comment by CookedApps** [comment]
How do i use this fix in gradle? I can't figure out how to download the latest commit by anubiann00b as jar repository.

**Comment by desertkun** [comment]
Just clone it and do `mvn compile install` to install it into your local maven repository. Also make sure to have `mavenLocal()` in your `build.gradle` script. Also, change `compile` to `com.esotericsoftware:kryonet:2.22.0-RC1` in `dependencies` section.

**Comment by CookedApps** [comment]
I have installed maven and cloned the patched kronet version to my local maven repo and added mavenLocal() and compile "com.esotericsoftware:kryonet:2.22.0-RC1" to my build.gradle..
But udp is not working though, i think it takes kryonet from mavenCentral because the complie command for that is the same. But i need to use mavenCentral also for other repos..
So how do make the build.gradle to use the repo on my local maven?
Thanks for your help desertkun!

**Comment by desertkun** [comment]
Just make sure to have `mavenLocal` before `mavenCentral`.

    mavenLocal()
    mavenCentral()

**Comment by CookedApps** [comment]
Yes, i have done it like that. So why isn't my Android 5 device receiving any udp messages?
I have to make an addition here: 
 ```java  
compile ("com.esotericsoftware:kryonet:2.22.0-RC1") {
         exclude module : 'kryo'
}
compile "com.esotericsoftware:kryo:3.0.1"```

This is how my compile looks like. I had to exclude kryo from kryonet and complie it seperately because if i only compile ```java "com.esotericsoftware:kryonet:2.22.0-RC1"``` i get an Multiple dex file exception.

**Comment by doomtoo** [comment]
Thanks also codeflakes0! Saved a lot of headache, and working for me on my newer Android devices now. Just switched to using UDP and TCP, and of course doesn't work on any but the oldest Android, but I was lucky to find this post.

Now if only I could figure out how to get UDP to register not based on initial IP... (when connecting from an external IP, it first gets the first message from the router IP, than the internal IP, so kryo discards every message after that because they aren't from a registered IP) :/

**Comment by Ash98** [comment]
For anyone who doesn't want to mess with maven, here is a .jar file of the patched version:

http://www57.zippyshare.com/v/w8MHHZXY/file.html

(Note you will have to compile Kryo seperately like shown above by CookedApps)

**Comment by simernes** [comment]
Thanks a lot @codeflakes0 , and the team for making this great project! This makes it work, however after implementing this fix I have an issue with connections taking a long time to establish, and I get a lot of warnings that look like this:

`Long monitor contention event with owner method=void com.esotericsoftware.kryonet.Client.connect(int, java.net.InetAddress, int, int) from Client.java:170 waiters=0 for 128ms`

`Long monitor contention event with owner method=int java.nio.SelectorImpl.selectInternal(long) from SelectorImpl.java:170 waiters=0 for 251ms`

`Long monitor contention event with owner method=int libcore.io.Posix.poll(android.system.StructPollfd[], int) from Posix.java:4294967294 waiters=0 for 248ms`

Any chance anyone else has faced this issue and come up with a solution?

**Comment by crykn** [comment]
The project itself wasn't updated for more than two years – have you tried using one of the newer forks?

**Comment by simernes** [comment]
You're right. No, I haven't, and since everything works with this one and I'm running my project on many different platforms and setups I was hoping I could stick with what I got implemented. Would you be able to recommend one though?

**Comment by crykn** [comment]
There are three main forks:
- The one by [Edarke](https://github.com/Edarke/kryonet), that adds support for queries (just see his [README](https://github.com/Edarke/kryonet/blob/master/README.md)), but removes support for a few of the bigger framework features like RMI
- The fork by [Pyeroh](https://github.com/Pyeroh/kryonet): This fork adds the most demanded bug fixes to kryonet and updates the kryo version to 4.0.0
- [My fork](https://github.com/crykn/kryonet): This fork is probably the [most actively maintained](https://github.com/EsotericSoftware/kryonet/network) right now and is based on Pyeroh's one. It adds a few more things from the kryonet bug tracker (listener is an interface, an improved documentation, a gradle based setup) and an easy to use type listener. For a complete list see the [README](https://github.com/crykn/kryonet/blob/master/README.md). Since this fork only adds minor fixes, it doesn't even require you to make changes to your code.

**Comment by simernes** [comment]
FYI I've been testing on different platforms, and the issue seems to have been resolved starting from Android 7.0, API 24. The patch that I have seen implemented for this issue (the one proposed by @codeflakes0 here) for me at least caused the time spent connecting to increase tremendously, which made my app look like it wasn't working.

**Comment by crykn** [comment]
If anyone is looking for a code base without the fix, my [fork](https://github.com/crykn/kryonet) allows you to disable it by setting `UdpConnection#androidFixDisabled`.
